onnxruntime/core/providers/nnapi/nnapi_builtin/model.cc

```diff
@@ -78,37 +78,48 @@ size_t Model::GetMappedOutputIdx(const std::string& name) const {
   return output_map_.at(name);
 }
 
-void Model::SetInputBuffer(const int32_t index, const InputBuffer& input) {
-  PrepareForExecution();
+Status Model::SetInputBuffer(const int32_t index, const InputBuffer& input) {
+  ORT_RETURN_IF_ERROR(PrepareForExecution());
 
-  THROW_ON_ERROR(nnapi_->ANeuralNetworksExecution_setInput(
+  RETURN_STATUS_ON_ERROR(nnapi_->ANeuralNetworksExecution_setInput(
       execution_, index, &input.type.operandType, input.buffer, input.type.GetOperandBlobByteSize()));
+
+  return Status::OK();
 }
 
-void Model::SetOutputBuffer(const int32_t index, const OutputBuffer& output) {
-  PrepareForExecution();
+Status Model::SetOutputBuffer(const int32_t index, const OutputBuffer& output) {
+  ORT_RETURN_IF_ERROR(PrepareForExecution());
 
   LOGS_DEFAULT(VERBOSE) << "Model::SetOutputBuffer, output shape "
                         << Shape2String(output.type.dimensions);
 
-  THROW_ON_ERROR(nnapi_->ANeuralNetworksExecution_setOutput(
+  RETURN_STATUS_ON_ERROR(nnapi_->ANeuralNetworksExecution_setOutput(
       execution_, index, &output.type.operandType, output.buffer, output.type.GetOperandBlobByteSize()));
+
+  return Status::OK();
+}
+
+bool Model::SupportsDynamicOutputShape() const {
+  // dynamic output shape is only supported on Android API levle 29+
+  return GetAndroidSdkVer() >= 29 && dynamic_output_buffer_size_ > 0;
 }
 
-void Model::PrepareForExecution() {
+Status Model::PrepareForExecution() {
   if (prepared_for_exe_)
-    return;
+    return Status::OK();
 
-  ORT_ENFORCE(nullptr != compilation_,
-              "Error in PrepareForExecution, compilation_ is null");
+  ORT_RETURN_IF_NOT(nullptr != compilation_,
+                    "Error in PrepareForExecution, compilation_ is null");
 
   // Copy the shaper for calculate the dynamic output shape
   // based on the input shape
   shaper_for_exeuction_ = shaper_;
 
```

Comment:
What's the reason for a copy being needed? Do we expect something else to change shaper_?

Reply:
The same model can be executed repeatedly with different size input, I would prefer a fresh copy for every execution


---
In reply to: [462109423](https://github.com/microsoft/onnxruntime/pull/4650#discussion_r462109423) [](ancestors = 462109423)